File: brick_nav/src/waypoint_publisher.cpp
#include <string>
#include <vector>

#include <iostream>
#include "ros/ros.h"
#include <geometry_msgs/PoseWithCovarianceStamped.h>
#include <cmath>
#include <time.h>
// #include <tf>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2_ros/transform_listener.h>
#include <geometry_msgs/TransformStamped.h>
#include <std_msgs/Empty.h>

static ros::Publisher pub_wp;
static bool publish = false;
static geometry_msgs::PoseWithCovarianceStamped waypoint;


void home_callback(const std_msgs::Empty msg) {

    waypoint.header.stamp = ros::Time::now();

    tf2::Quaternion quaternion;
    quaternion.setRPY(0, 0, 0);

    waypoint.pose.pose.orientation.x = quaternion[0];
    waypoint.pose.pose.orientation.y = quaternion[1];
    waypoint.pose.pose.orientation.z = quaternion[2];
    waypoint.pose.pose.orientation.w = quaternion[3];

    waypoint.pose.pose.position.x = 0;
    waypoint.pose.pose.position.y = 0;

    publish = true;

}

void target_callback(const std_msgs::Empty msg){

    // Get brick location from tf tree
    tf2_ros::Buffer tfBuffer;
    tf2_ros::TransformListener tfListener(tfBuffer);
    geometry_msgs::TransformStamped transformStamped;

    while (ros::ok()){
        
        try{
            transformStamped = tfBuffer.lookupTransform("map", "target",ros::Time(0));
            break;
        }
        catch (tf2::TransformException &ex) {
            ROS_WARN("%s",ex.what());
            ros::Duration(1.0).sleep();
            continue;
        }
    }

    // Set up waypoint
    waypoint.header.stamp = ros::Time::now();

    tf2::Quaternion quaternion;
    quaternion.setRPY(0, 0, M_PI_2);

    waypoint.pose.pose.orientation.x = quaternion[0];
    waypoint.pose.pose.orientation.y = quaternion[1];
    waypoint.pose.pose.orientation.z = quaternion[2];
    waypoint.pose.pose.orientation.w = quaternion[3];

    waypoint.pose.pose.position.x = transformStamped.transform.translation.x;
    waypoint.pose.pose.position.y = transformStamped.transform.translation.y;

    publish = true;

}

int main(int argc, char** argv) {

    ros::init(argc, argv, "waypoint_publisher");
    ros::NodeHandle nh;

    // Set up publishers and subscribers

    pub_wp = nh.advertise<geometry_msgs::PoseWithCovarianceStamped>("/my_jackal_waypoints", 1);
    ros::Publisher pub_init_wp = nh.advertise<std_msgs::Empty>("/path_ready", 1);

    ros::Subscriber home_sub = nh.subscribe<std_msgs::Empty>(
        "/home_ready", 1, home_callback);
    ros::Subscriber target_sub = nh.subscribe<std_msgs::Empty>(
        "/target_ready", 1, target_callback);

    waypoint.header.frame_id = "/map";

    // Establish loop rate
    ros::Rate r(10);

    while (ros::ok()){
        
        if (publish){

            // Publish waypoint
            while (ros::ok()){

                int connections = pub_wp.getNumSubscribers();
                if (connections > 0){
                    pub_wp.publish(waypoint);
                    break;
                }
                std::cout << "Waiting for /my_jackal_waypoints topic" << std::endl;
            }

            std::cout << "Published waypoint" << std::endl;

            // Publish Waypoint follower start command
            std_msgs::Empty start_command;
    
            while (ros::ok()){
                int connections = pub_init_wp.getNumSubscribers();
                if (connections > 0){
                    pub_init_wp.publish(start_command);
                    std::cout << "Sent waypoint execution command" << std::endl;
                    break;
                }

                std::cout << "Waiting for /path_ready topic" << std::endl;
            }

            publish = false;
        }

        r.sleep();
        ros::spinOnce();
    }

    return 0;
}

File: brick_nav/src/waypoint_publisher.py
#!/usr/bin/env python
# license removed for brevity
import rospy
from std_msgs.msg import String
from geometry_msgs.msg import PoseWithCovarianceStamped
import math
import time
import tf
import tf2_ros
from std_msgs.msg import Empty

"""
user:~$ rosmsg show geometry_msgs/PoseWithCovarianceStamped                                                                                                               
std_msgs/Header header                                                                                                                                                    
  uint32 seq                                                                                                                                                              
  time stamp                                                                                                                                                              
  string frame_id                                                                                                                                                         
geometry_msgs/PoseWithCovariance pose                                                                                                                                     
  geometry_msgs/Pose pose                                                                                                                                                 
    geometry_msgs/Point position                                                                                                                                          
      float64 x                                                                                                                                                           
      float64 y                                                                                                                                                           
      float64 z                                                                                                                                                           
    geometry_msgs/Quaternion orientation                                                                                                                                  
      float64 x                                                                                                                                                           
      float64 y                                                                                                                                                           
      float64 z                                                                                                                                                           
      float64 w                                                                                                                                                           
  float64[36] covariance 
"""

def talker():

    pub_wp = rospy.Publisher('/my_jackal_waypoints', PoseWithCovarianceStamped, queue_size=1)
    
    pub_init_wp = rospy.Publisher('/path_ready', Empty, queue_size=1)
    
    rospy.init_node('waypoint_publisher', anonymous=True)
    rate = rospy.Rate(10) # 10hz
    
    tfBuffer = tf2_ros.Buffer()
    listener = tf2_ros.TransformListener(tfBuffer)

    while(True):
      try:
          trans = tfBuffer.lookup_transform('map', 'cluster_1_frame', rospy.Time())
          break
      except (tf2_ros.LookupException, tf2_ros.ConnectivityException, tf2_ros.ExtrapolationException):
          pass

    my_wp = PoseWithCovarianceStamped()
    my_wp.header.stamp = rospy.Time.now()
    my_wp.header.frame_id = "/map"
    
    
    # init_value = 2.0
    
    roll = 0
    pitch = 0
    yaw = math.pi/2
    quaternion = tf.transformations.quaternion_from_euler(roll, pitch, yaw)
    #type(pose) = geometry_msgs.msg.Pose
    my_wp.pose.pose.orientation.x = quaternion[0]
    my_wp.pose.pose.orientation.y = quaternion[1]
    my_wp.pose.pose.orientation.z = quaternion[2]
    my_wp.pose.pose.orientation.w = quaternion[3]

    print(trans.transform.translation.x,trans.transform.translation.y)

    for i in range(1):
        rospy.loginfo(str(i))
        my_wp.pose.pose.position.x = float(trans.transform.translation.x) #+ init_value 
        my_wp.pose.pose.position.y = float(trans.transform.translation.y) #+ init_value
        
        while not rospy.is_shutdown():
            connections = pub_wp.get_num_connections()
            if connections > 0:
                pub_wp.publish(my_wp)
                break
            rospy.loginfo("Waiting for /my_jackal_waypoints topic")
            rate.sleep()
        rospy.loginfo("Published waypoint number ="+str(i))
        time.sleep(2)
    
    start_command = Empty()
    
    while not rospy.is_shutdown():
      connections = pub_init_wp.get_num_connections()
      if connections > 0:
          pub_init_wp.publish(start_command)
          rospy.loginfo("Sent waypoint list execution command")
          break
      rospy.loginfo("Waiting for /path_ready topic")
      rate.sleep()


if __name__ == '__main__':

    

    try:
        talker()
    except rospy.ROSInterruptException:
        pass